crates/dslab-core/src/state.rs

```diff
@@ -74,9 +76,52 @@ impl SimulationState {
         }
     }
 
+    pub fn add_ordered_event<T>(&mut self, data: T, src: Id, dest: Id, delay: f64) -> EventId
+    where
+        T: EventData,
+    {
+        if !self.can_add_ordered_event(delay) {
+            panic!("Event order is broken! Make sure you are passing events to emit_*_ordered functions in non-decreasing order of time.");
+        }
+        let event_id = self.event_count;
+        let event = Event {
+            id: event_id,
+            time: self.clock + delay,
+            src,
+            dest,
+            data: Box::new(data),
+        };
+        if delay >= 0. {
+            self.ordered_events.push_back(event);
+            self.event_count += 1;
+            event_id
+        } else {
+            log_incorrect_event(event, &format!("negative delay {}", delay));
+            panic!("Event delay is negative! It is not allowed to add events from the past.");
+        }
+    }
+
+    pub fn can_add_ordered_event(&self, delay: f64) -> bool {
+        if let Some(evt) = self.ordered_events.back() {
+            if delay + self.clock < evt.time - 1e-12 {
+                return false;
+            }
+        }
+        true
```

Comment:
Предлагаю убрать `1e-12` и возложить решение возможных проблем с точностью на пользователя.

```suggestion
        self.ordered_events
            .back()
            .map(|last| last.time <= self.clock + delay)
            .unwrap_or(true)
```

Reply:
Я опасаюсь, что тут может возникнуть очень маленькая ошибка в delay + self.clock, играющая роль при delay + self.clock == evt.time, причем погрешность вызвана не действиями пользователя.

Reply:
Если будет ошибка, то пользователь пусть решает проблему, например путем округления времен событий или использования max(clock+delay, prev_event_time).

Reply:
У тебя самого такая ошибка была?

Reply:
Не знаю, я сразу такое условие сделал. Но у меня были ошибки с добавлением delay порядка -1e-16 (фактически ноль), и наличие таких ошибок не всегда очевидно

Reply:
Тогда надо не только в can_add_ordered_event это учитывать, но и когда добавляем событие у которого время меньше clock, то округлять вверх до clock, а то монотонности в симуляции не будет.

Reply:
Ну и обязательно про это написать в документации

Reply:
Округлять пожалуй долго будет, а события с временем меньше clock вроде как паникуют при добавлении. Или ты предлагаешь заменять delay на max(delay, 0) для delay > -epsilon?

Reply:
Допустим clock=90, last=100 (последнее время в буфере). Приходит запрос add_ordered_event c delay 9.99.... Допустим проверка в can_add_ordered_event пройдена, тогда событие будет добавлено в буфер c временем 99.99... Паники не будет, но теперь события в буфере не упорядочены по времени.

Reply:
Да, действительно. Сделаю так.

Reply:
Сделал

Reply:
Кажется все еще остается проблема с отрицательным delay. Допустим clock=100, last=100 (последнее время в буфере). Приходит запрос add_ordered_event c delay -0.000.... Допустим проверка в can_add_ordered_event пройдена, тогда событие будет добавлено в буфер c временем 100 (работает max). Но потом будет паника из-за отрицательного delay.

Reply:
Так если паника будет, то неважно, было ли событие добавлено или нет

Reply:
Так я думал мы такие события хотим добавлять, и не паниковать если очень маленький отрицательный delay

Reply:
Нет, я хотел избавиться от проблем, когда в условии <= возникает равенство, но из-за погрешности сложения в левой части условие нарушается. Если пытаться решить проблему с delay, то это надо ещё и при добавлении в кучу epsilon прибавлять